csharp/ql/src/semmle/code/csharp/dataflow/internal/rangeanalysis/ModulusAnalysisSpecific.qll

```diff
@@ -0,0 +1,99 @@
+module Private {
+  private import csharp as CS
+  private import ConstantUtils as CU
+  private import semmle.code.csharp.controlflow.Guards as G
+  private import semmle.code.csharp.dataflow.internal.rangeanalysis.RangeUtils as RU
+  private import SsaUtils as SU
+  private import SignAnalysisSpecific::Private as SA
+
+  class BasicBlock = CS::Ssa::BasicBlock;
+
+  class SsaVariable extends CS::Ssa::Definition {
+    CS::AssignableRead getAUse() { result = this.getARead() }
+  }
+
+  class SsaPhiNode = CS::Ssa::PhiNode;
+
+  class Expr = CS::Expr;
+
+  class Guard = G::Guard;
+
+  class ConstantIntegerExpr = CU::ConstantIntegerExpr;
+
+  class ConditionalExpr extends CS::ConditionalExpr {
+    /** Gets the "then" expression of this conditional expression. */
+    Expr getTrueExpr() { result = this.getThen() }
+
+    /** Gets the "else" expression of this conditional expression. */
+    Expr getFalseExpr() { result = this.getElse() }
+  }
+
+  class AddExpr = CS::AddExpr;
+
+  class SubExpr = CS::SubExpr;
+
+  class RemExpr = CS::RemExpr;
+
+  class AssignAddExpr extends CS::AssignAddExpr {
+    /** Gets the left operand of this assignment. */
+    Expr getDest() { result = this.getLValue() }
+
+    /** Gets the right operand of this assignment. */
+    Expr getRhs() { result = this.getRValue() }
+  }
+
+  class AssignSubExpr extends CS::AssignSubExpr {
+    /** Gets the left operand of this assignment. */
+    Expr getDest() { result = this.getLValue() }
+
+    /** Gets the right operand of this assignment. */
+    Expr getRhs() { result = this.getRValue() }
+  }
+
+  class AndBitwiseExpr extends CS::BitwiseAndExpr {
+    predicate hasOperands(Expr e1, Expr e2) {
+      (
+        this.getAnOperand() = e1 or
+        this.getAnOperand() = e2
+      ) and
+      e1 != e2
+    }
+  }
+
+  class AssignAndExpr extends CS::AssignAndExpr {
+    /** Gets the an operand of this assignment. */
+    Expr getSource() { result = this.getRValue() or result = this.getLValue() }
+  }
+
+  class MulExpr = CS::MulExpr;
+
+  class AssignMulExpr extends CS::AssignMulExpr {
+    /** Gets the an operand of this assignment. */
+    Expr getSource() { result = this.getRValue() or result = this.getLValue() }
+  }
+
+  class LShiftExpr = CS::LShiftExpr;
+
+  class AssignLShiftExpr extends CS::AssignLShiftExpr {
+    /** Gets the right operand of this assignment. */
+    Expr getRhs() { result = this.getRValue() }
+  }
```

Comment:
I think this interface can be simplified. For the `+` case (the other cases are similar), we could for C# define
```ql
class AddExpr extends CS::AddExpr {
  Expr getLhs() { result = this.getLeftOperand() }
  Expr getRhs() { result = this.getRightOperand() }
}
```
remove `AssignAddExpr`, and then in `ModulusAnalysisCommon.qll` we could join all cases involving `AddExpr` and `AssignAddExpr` to just a single case using `AddExpr`. (Note that this only works provided that we add `valueFlowStep(e2.(AssignOperation).getExpandedAssignment(), e1, delta)` to the `valueStep` predicate, see suggestion further down).

For Java, it would be
```ql
class AddExpr extends J::Expr {
    AddExpr() { this instanceof J::AddExpr or this instanceof J::AssignAddExpr }

    Expr getLhs() {
      result = this.(J::AddExpr).getLeftOperand()
      or
      result = this.(J::AssignAddExpr).getDest()
    }

    Expr getRhs() {
      result = this.(J::AddExpr).getRightOperand()
      or
      result = this.(J::AssignAddExpr).getRhs()
    }
  }
```
The same pattern for `-`, `*`, `<<`, and `&`.